If jetty.org is down, impossible to start Jetty server (12.0.5) - multiple dtd to eclipse and jetty

**Jetty version(s)**
jetty-12.0.5

**Jetty Environment**
ee10

**Java version/vendor** `(use: java -version)`
openjdk 17.0.9 2023-10-17

**OS type/version**
Linux or Windows

**Description**
xml in webapps, like
```
<?xml version="1.0"  encoding="UTF-8"?>
<!DOCTYPE Configure PUBLIC "-//Mort Bay Consulting//DTD Configure//EN" "https://jetty.org/configure_10_0.dtd">
<Configure class="org.eclipse.jetty.server.handler.ContextHandler">
...
```

While https://jetty.org/ is down, I got error

```
2024-10-18 10:41:07.163 [main] WARN  o.e.jetty.deploy.DeploymentManager - Unable to reach node goal: started
java.net.SocketException: Unexpected end of file from server
	at java.base/sun.net.www.http.HttpClient.parseHTTPHeader(HttpClient.java:954)
	at java.base/sun.net.www.http.HttpClient.parseHTTP(HttpClient.java:761)
	at java.base/sun.net.www.http.HttpClient.parseHTTPHeader(HttpClient.java:951)
	at java.base/sun.net.www.http.HttpClient.parseHTTP(HttpClient.java:761)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1688)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1589)
	at java.base/sun.net.www.protocol.https.HttpsURLConnectionImpl.getInputStream(HttpsURLConnectionImpl.java:224)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLEntityManager.setupCurrentEntity(XMLEntityManager.java:677)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLEntityManager.startEntity(XMLEntityManager.java:1397)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLEntityManager.startDTDEntity(XMLEntityManager.java:1363)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDTDScannerImpl.setInputSource(XMLDTDScannerImpl.java:257)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$DTDDriver.dispatch(XMLDocumentScannerImpl.java:1152)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$DTDDriver.next(XMLDocumentScannerImpl.java:1040)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$PrologDriver.next(XMLDocumentScannerImpl.java:943)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:605)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.next(XMLNSDocumentScannerImpl.java:112)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:542)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:889)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:825)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1224)
	at java.xml/com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl$JAXPSAXParser.parse(SAXParserImpl.java:637)
	at java.xml/com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl.parse(SAXParserImpl.java:326)
	at org.eclipse.jetty.xml.XmlParser.parse(XmlParser.java:254)
	at org.eclipse.jetty.xml.XmlParser.parse(XmlParser.java:303)
	at org.eclipse.jetty.xml.XmlConfiguration.<init>(XmlConfiguration.java:278)
	at org.eclipse.jetty.deploy.providers.ContextProvider$1.<init>(ContextProvider.java:350)
	at org.eclipse.jetty.deploy.providers.ContextProvider.createContextHandler(ContextProvider.java:349)
	at org.eclipse.jetty.deploy.App.getContextHandler(App.java:108)
	at org.eclipse.jetty.deploy.bindings.StandardDeployer.processBinding(StandardDeployer.java:33)
	at org.eclipse.jetty.deploy.AppLifeCycle.runBindings(AppLifeCycle.java:183)
	at org.eclipse.jetty.deploy.DeploymentManager.requestAppGoal(DeploymentManager.java:522)
	at org.eclipse.jetty.deploy.DeploymentManager.addApp(DeploymentManager.java:162)
	at org.eclipse.jetty.deploy.providers.ScanningAppProvider.pathAdded(ScanningAppProvider.java:293)
	at org.eclipse.jetty.deploy.providers.ContextProvider.pathAdded(ContextProvider.java:520)
	at org.eclipse.jetty.deploy.providers.ScanningAppProvider$1.pathAdded(ScanningAppProvider.java:69)
	at org.eclipse.jetty.util.Scanner.reportAddition(Scanner.java:902)
	at org.eclipse.jetty.util.Scanner.reportDifferences(Scanner.java:868)
	at org.eclipse.jetty.util.Scanner.scan(Scanner.java:775)
	at org.eclipse.jetty.util.Scanner.startScanning(Scanner.java:664)
	at org.eclipse.jetty.util.Scanner.doStart(Scanner.java:636)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:93)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:171)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:121)
	at org.eclipse.jetty.deploy.providers.ScanningAppProvider.doStart(ScanningAppProvider.java:264)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:93)
	at org.eclipse.jetty.deploy.DeploymentManager.startAppProvider(DeploymentManager.java:592)
	at org.eclipse.jetty.deploy.DeploymentManager.doStart(DeploymentManager.java:263)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:93)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:171)
	at org.eclipse.jetty.server.Server.start(Server.java:622)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:121)
	at org.eclipse.jetty.server.Handler$Abstract.doStart(Handler.java:468)
	at org.eclipse.jetty.server.Server.doStart(Server.java:563)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:93)
	at org.eclipse.jetty.xml.XmlConfiguration.main(XmlConfiguration.java:2030)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.eclipse.jetty.start.Main.invokeMain(Main.java:221)
	at org.eclipse.jetty.start.Main.start(Main.java:519)
	at org.eclipse.jetty.start.Main.main(Main.java:76)
2024-10-18 10:41:07.184 [main] WARN  o.eclipse.jetty.xml.XmlConfiguration - Unable to execute XmlConfiguration
java.net.SocketException: Unexpected end of file from server
	at java.base/sun.net.www.http.HttpClient.parseHTTPHeader(HttpClient.java:954)
	at java.base/sun.net.www.http.HttpClient.parseHTTP(HttpClient.java:761)
	at java.base/sun.net.www.http.HttpClient.parseHTTPHeader(HttpClient.java:951)
	at java.base/sun.net.www.http.HttpClient.parseHTTP(HttpClient.java:761)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1688)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1589)
	at java.base/sun.net.www.protocol.https.HttpsURLConnectionImpl.getInputStream(HttpsURLConnectionImpl.java:224)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLEntityManager.setupCurrentEntity(XMLEntityManager.java:677)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLEntityManager.startEntity(XMLEntityManager.java:1397)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLEntityManager.startDTDEntity(XMLEntityManager.java:1363)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDTDScannerImpl.setInputSource(XMLDTDScannerImpl.java:257)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$DTDDriver.dispatch(XMLDocumentScannerImpl.java:1152)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$DTDDriver.next(XMLDocumentScannerImpl.java:1040)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$PrologDriver.next(XMLDocumentScannerImpl.java:943)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:605)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.next(XMLNSDocumentScannerImpl.java:112)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:542)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:889)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:825)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1224)
	at java.xml/com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl$JAXPSAXParser.parse(SAXParserImpl.java:637)
	at java.xml/com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl.parse(SAXParserImpl.java:326)
	at org.eclipse.jetty.xml.XmlParser.parse(XmlParser.java:254)
	at org.eclipse.jetty.xml.XmlParser.parse(XmlParser.java:303)
	at org.eclipse.jetty.xml.XmlConfiguration.<init>(XmlConfiguration.java:278)
	at org.eclipse.jetty.deploy.providers.ContextProvider$1.<init>(ContextProvider.java:350)
	at org.eclipse.jetty.deploy.providers.ContextProvider.createContextHandler(ContextProvider.java:349)
	at org.eclipse.jetty.deploy.App.getContextHandler(App.java:108)
	at org.eclipse.jetty.deploy.bindings.StandardDeployer.processBinding(StandardDeployer.java:33)
	at org.eclipse.jetty.deploy.AppLifeCycle.runBindings(AppLifeCycle.java:183)
	at org.eclipse.jetty.deploy.DeploymentManager.requestAppGoal(DeploymentManager.java:522)
	at org.eclipse.jetty.deploy.DeploymentManager.addApp(DeploymentManager.java:162)
	at org.eclipse.jetty.deploy.providers.ScanningAppProvider.pathAdded(ScanningAppProvider.java:293)
	at org.eclipse.jetty.deploy.providers.ContextProvider.pathAdded(ContextProvider.java:520)
	at org.eclipse.jetty.deploy.providers.ScanningAppProvider$1.pathAdded(ScanningAppProvider.java:69)
	at org.eclipse.jetty.util.Scanner.reportAddition(Scanner.java:902)
	at org.eclipse.jetty.util.Scanner.reportDifferences(Scanner.java:868)
	at org.eclipse.jetty.util.Scanner.scan(Scanner.java:775)
	at org.eclipse.jetty.util.Scanner.startScanning(Scanner.java:664)
	at org.eclipse.jetty.util.Scanner.doStart(Scanner.java:636)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:93)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:171)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:121)
	at org.eclipse.jetty.deploy.providers.ScanningAppProvider.doStart(ScanningAppProvider.java:264)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:93)
	at org.eclipse.jetty.deploy.DeploymentManager.startAppProvider(DeploymentManager.java:592)
	at org.eclipse.jetty.deploy.DeploymentManager.doStart(DeploymentManager.java:263)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:93)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:171)
	at org.eclipse.jetty.server.Server.start(Server.java:622)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:121)
	at org.eclipse.jetty.server.Handler$Abstract.doStart(Handler.java:468)
	at org.eclipse.jetty.server.Server.doStart(Server.java:563)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:93)
	at org.eclipse.jetty.xml.XmlConfiguration.main(XmlConfiguration.java:2030)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.eclipse.jetty.start.Main.invokeMain(Main.java:221)
	at org.eclipse.jetty.start.Main.start(Main.java:519)
	at org.eclipse.jetty.start.Main.main(Main.java:76)
```


**How to reproduce?**
Shutdown jetty.org ? ;)

About the test with system's hosts and jetty.org to a false IP

I tried, and yes, it's reproduced: but with another error **java.net.UnknownHostException: jetty.org** with 0.0.0.0

```
java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at org.eclipse.jetty.start.Main.invokeMain(Main.java:221)
        at org.eclipse.jetty.start.Main.start(Main.java:519)
        at org.eclipse.jetty.start.Main.main(Main.java:76)
Caused by: java.net.UnknownHostException: jetty.org
        at java.base/sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:572)
        at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:327)
        at java.base/java.net.Socket.connect(Socket.java:633)
...
```

With 192.168.88.99: Connection timed out
```
java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at org.eclipse.jetty.start.Main.invokeMain(Main.java:221)
        at org.eclipse.jetty.start.Main.start(Main.java:519)
        at org.eclipse.jetty.start.Main.main(Main.java:76)
Caused by: java.net.ConnectException: Connection timed out: connect
        at java.base/sun.nio.ch.Net.connect0(Native Method)
...
```

--- a/documentation/jetty/modules/operations-guide/pages/protocols/index.adoc
+++ b/documentation/jetty/modules/operations-guide/pages/protocols/index.adoc
@@ -616,7 +616,7 @@ Use the following file as example, copy it as `$JETTY_BASE/webapps/wordpress.xml
 [,xml,options=nowrap]
 ----
 <?xml version="1.0"  encoding="UTF-8"?>
-<!DOCTYPE Configure PUBLIC "-//Mort Bay Consulting//DTD Configure//EN" "https://jetty.org/configure_10_0.dtd">
+<!DOCTYPE Configure PUBLIC "-//Jetty//Configure//EN" "https://jetty.org/configure_10_0.dtd">
 <Configure class="org.eclipse.jetty.server.handler.ContextHandler">
 
   <New id="root" class="java.lang.String">

--- a/jetty-core/jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java
+++ b/jetty-core/jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java
@@ -51,6 +51,7 @@ import java.util.ServiceLoader;
 import java.util.Set;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
+import javax.xml.XMLConstants;
 import javax.xml.parsers.SAXParserFactory;
 
 import org.eclipse.jetty.util.ConcurrentPool;
@@ -2074,6 +2075,10 @@ public class XmlConfiguration
         {
             // Use JVM default implementation (not the one found in the classloader from non-JVM sources)
             SAXParserFactory factory = SAXParserFactory.newDefaultInstance();
+            // Use secure processing factory level defaults (to allow
+            // newly created SAXParsers from this factory to be initialized properly
+            // for external entity handling)
+            setFeature(factory, XMLConstants.FEATURE_SECURE_PROCESSING, true);
             // Don't allow use of XInclude to reference external entities.
             factory.setXIncludeAware(false);
             return factory;
@@ -2082,6 +2087,7 @@ public class XmlConfiguration
         @Override
         public void setValidating(boolean validating)
         {
+            // Initialize the SAXParser and generic XmlParser behaviors.
             super.setValidating(validating);
 
             try
@@ -2089,9 +2095,18 @@ public class XmlConfiguration
                 XMLReader xmlReader = getSAXParser().getXMLReader();
 
                 // disable all external entity references with Jetty's Configuration XML.
-                xmlReader.setFeature("http://xml.org/sax/features/external-general-entities", false);
-                xmlReader.setFeature("http://xml.org/sax/features/external-parameter-entities", false);
-                xmlReader.setFeature("http://apache.org/xml/features/nonvalidating/load-external-dtd", false);
+
+                // These features configure the XMLEntityManager that the SAXParser
+                // (and Xerces) uses.   These features are applied to the current
+                // entity (the document being parsed) and how the referenced entities
+                // are to be looked up.  The resulting XMLEntityManager is configured
+                // per document being parsed.
+
+                // Configure SAX
+                setFeature(xmlReader, "http://xml.org/sax/features/external-general-entities", false);
+                setFeature(xmlReader, "http://xml.org/sax/features/external-parameter-entities", false);
+                // Configure Xerces
+                setFeature(xmlReader, "http://apache.org/xml/features/nonvalidating/load-external-dtd", false);
             }
             catch (SAXException e)
             {

--- a/jetty-core/jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java
+++ b/jetty-core/jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java
@@ -52,6 +52,7 @@ import java.util.Set;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 import javax.xml.XMLConstants;
+import javax.xml.parsers.SAXParser;
 import javax.xml.parsers.SAXParserFactory;
 
 import org.eclipse.jetty.util.ConcurrentPool;
@@ -2085,11 +2086,8 @@ public class XmlConfiguration
         }
 
         @Override
-        public void setValidating(boolean validating)
+        protected void configure(SAXParser saxParser)
         {
-            // Initialize the SAXParser and generic XmlParser behaviors.
-            super.setValidating(validating);
-
             try
             {
                 XMLReader xmlReader = getSAXParser().getXMLReader();

--- a/jetty-core/jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java
+++ b/jetty-core/jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java
@@ -51,6 +51,7 @@ import java.util.ServiceLoader;
 import java.util.Set;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
+import javax.xml.parsers.SAXParserFactory;
 
 import org.eclipse.jetty.util.ConcurrentPool;
 import org.eclipse.jetty.util.ExceptionUtil;
@@ -69,6 +70,7 @@ import org.eclipse.jetty.util.resource.ResourceFactory;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.xml.sax.SAXException;
+import org.xml.sax.XMLReader;
 
 /**
  * <p>Configures objects from XML.</p>
@@ -2066,6 +2068,35 @@ public class XmlConfiguration
             if (catalogUrl == null)
                 throw new IllegalStateException("Catalog not found: catalog-configure.xml");
             addCatalog(URI.create(catalogUrl.toExternalForm()));
+        }
+
+        protected SAXParserFactory newSAXParserFactory()
+        {
+            // Use JVM default implementation (not the one found in the classloader from non-JVM sources)
+            SAXParserFactory factory = SAXParserFactory.newDefaultInstance();
+            // Don't allow use of XInclude to reference external entities.
+            factory.setXIncludeAware(false);
+            return factory;
+        }
+
+        @Override
+        public void setValidating(boolean validating)
+        {
+            super.setValidating(validating);
+
+            try
+            {
+                XMLReader xmlReader = getSAXParser().getXMLReader();
+
+                // disable all external entity references with Jetty's Configuration XML.
+                xmlReader.setFeature("http://xml.org/sax/features/external-general-entities", false);
+                xmlReader.setFeature("http://xml.org/sax/features/external-parameter-entities", false);
+                xmlReader.setFeature("http://apache.org/xml/features/nonvalidating/load-external-dtd", false);
+            }
+            catch (SAXException e)
+            {
+                LOG.warn(e.getMessage());
+            }
         }
 
         @Override

--- a/jetty-core/jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlParser.java
+++ b/jetty-core/jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlParser.java
@@ -112,7 +112,7 @@ public class XmlParser
         return SAXParserFactory.newInstance();
     }
 
-    protected void setFeature(SAXParserFactory factory, String name, boolean value)
+    protected static void setFeature(SAXParserFactory factory, String name, boolean value)
     {
         try
         {
@@ -124,7 +124,7 @@ public class XmlParser
         }
     }
 
-    protected void setFeature(XMLReader xmlReader, String name, boolean value)
+    protected static void setFeature(XMLReader xmlReader, String name, boolean value)
     {
         try
         {

--- a/jetty-core/jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlParser.java
+++ b/jetty-core/jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlParser.java
@@ -69,6 +69,7 @@ public class XmlParser
     private Object _xpaths;
     private String _dtd;
     private List<EntityResolver> _entityResolvers = new ArrayList<>();
+    private SAXParserFactory factory;
 
     /**
      * Construct XmlParser
@@ -112,6 +113,16 @@ public class XmlParser
         return SAXParserFactory.newInstance();
     }
 
+    protected SAXParser newSAXParser() throws ParserConfigurationException, SAXException
+    {
+        return factory.newSAXParser();
+    }
+
+    protected void configure(SAXParser saxParser)
+    {
+        /* override to configure sax parser at the right time */
+    }
+
     protected static void setFeature(SAXParserFactory factory, String name, boolean value)
     {
         try
@@ -140,9 +151,9 @@ public class XmlParser
     {
         try
         {
-            SAXParserFactory factory = newSAXParserFactory();
+            factory = newSAXParserFactory();
             factory.setValidating(validating);
-            _parser = factory.newSAXParser();
+            _parser = newSAXParser();
 
             try
             {
@@ -174,6 +185,10 @@ public class XmlParser
         {
             LOG.warn("Unable to set validating on XML Parser", e);
             throw new Error(e.toString());
+        }
+        finally
+        {
+            configure(_parser);
         }
     }
 

--- a/jetty-core/jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlParser.java
+++ b/jetty-core/jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlParser.java
@@ -31,6 +31,7 @@ import javax.xml.catalog.Catalog;
 import javax.xml.catalog.CatalogFeatures;
 import javax.xml.catalog.CatalogManager;
 import javax.xml.catalog.CatalogResolver;
+import javax.xml.parsers.ParserConfigurationException;
 import javax.xml.parsers.SAXParser;
 import javax.xml.parsers.SAXParserFactory;
 
@@ -43,6 +44,8 @@ import org.xml.sax.ContentHandler;
 import org.xml.sax.EntityResolver;
 import org.xml.sax.InputSource;
 import org.xml.sax.SAXException;
+import org.xml.sax.SAXNotRecognizedException;
+import org.xml.sax.SAXNotSupportedException;
 import org.xml.sax.SAXParseException;
 import org.xml.sax.XMLReader;
 import org.xml.sax.helpers.DefaultHandler;
@@ -107,6 +110,30 @@ public class XmlParser
     protected SAXParserFactory newSAXParserFactory()
     {
         return SAXParserFactory.newInstance();
+    }
+
+    protected void setFeature(SAXParserFactory factory, String name, boolean value)
+    {
+        try
+        {
+            factory.setFeature(name, value);
+        }
+        catch (SAXNotSupportedException | SAXNotRecognizedException | ParserConfigurationException e)
+        {
+            LOG.warn("Unable to setFeature({}, {})", name, value, e);
+        }
+    }
+
+    protected void setFeature(XMLReader xmlReader, String name, boolean value)
+    {
+        try
+        {
+            xmlReader.setFeature(name, value);
+        }
+        catch (SAXNotSupportedException | SAXNotRecognizedException e)
+        {
+            LOG.warn("Unable to setFeature({}, {})", name, value, e);
+        }
     }
 
     public void setValidating(boolean validating)

--- a/jetty-core/jetty-xml/src/test/java/org/eclipse/jetty/xml/XmlConfigurationTest.java
+++ b/jetty-core/jetty-xml/src/test/java/org/eclipse/jetty/xml/XmlConfigurationTest.java
@@ -1994,16 +1994,19 @@ public class XmlConfigurationTest
         assertThat(tc.testObject, instanceOf(ScheduledExecutorService.class));
     }
 
-    public static Stream<Arguments> xmlSystemIdSource()
+    private static List<String> systemIdVariants()
     {
-        List<Arguments> ids = new ArrayList<>();
-
-        String[] schemes = {"http", "https"};
-        String[] contexts = {"eclipse.org/jetty", "www.eclipse.org/jetty", "eclipse.dev/jetty", "www.eclipse.dev/jetty", "jetty.org"};
-        String[] paths = {"configure.dtd", // version-less
+        List<String> ids = new ArrayList<>();
+        List<String> schemes = List.of("http", "https");
+        List<String> contexts = List.of("eclipse.org/jetty",
+            "www.eclipse.org/jetty",
+            "eclipse.dev/jetty",
+            "www.eclipse.dev/jetty",
+            "jetty.org");
+        List<String> paths = List.of("configure.dtd", // version-less
                           "configure_9_0.dtd", // 9.0
                           "configure_9_3.dtd", // 9.3
-                          "configure_10_0.dtd"}; // 10.
+                          "configure_10_0.dtd"); // 10.0
 
         for (String scheme: schemes)
         {
@@ -2011,12 +2014,16 @@ public class XmlConfigurationTest
             {
                 for (String path: paths)
                 {
-                    ids.add(Arguments.of(String.format("%s://%s/%s", scheme, host, path)));
+                    ids.add(String.format("%s://%s/%s", scheme, host, path));
                 }
             }
         }
+        return ids;
+    }
 
-        return ids.stream();
+    public static Stream<Arguments> xmlSystemIdSource()
+    {
+        return systemIdVariants().stream().map(Arguments::of);
     }
 
     /**
@@ -2032,6 +2039,39 @@ public class XmlConfigurationTest
         XmlParser configurationProcessor = xmlConfiguration.getXmlParser();
         InputSource inputSource = configurationProcessor.resolveEntity(null, xmlSystemId);
         assertNotNull(inputSource, "SystemID: " + xmlSystemId + " does not exist");
+    }
+
+    public static Stream<Arguments> xmlPublicIdSource()
+    {
+        List<Arguments> cases = new ArrayList<>();
+
+        // XML Public IDs seen in use with Jetty.
+        List<String> publicIds = List.of("-//Jetty//Configure//EN",
+            "-//Mort Bay Consulting//DTD Configure//EN");
+
+        for (String systemId: systemIdVariants())
+        {
+            for (String publicId: publicIds)
+            {
+                cases.add(Arguments.of(publicId, systemId));
+            }
+        }
+        return cases.stream();
+    }
+
+    /**
+     * Test to ensure that all required XML PUBLIC ID variants are covered in the
+     * {@link XmlConfiguration} internals.
+     */
+    @ParameterizedTest
+    @MethodSource("xmlPublicIdSource")
+    public void testPublicIdVariants(String xmlPublicId, String xmlSystemId) throws IOException, SAXException
+    {
+        // empty raw xml, just to instantiate XmlConfiguration, so we can access the XmlParser / ConfigurationParser.
+        XmlConfiguration xmlConfiguration = asXmlConfiguration("<Configure class=\"org.eclipse.jetty.xml.TestConfiguration\" />");
+        XmlParser configurationProcessor = xmlConfiguration.getXmlParser();
+        InputSource inputSource = configurationProcessor.resolveEntity(xmlPublicId, xmlSystemId);
+        assertNotNull(inputSource, "reference: PUBLIC " + xmlPublicId + " SYSTEM " + xmlSystemId + " does not exist");
     }
 
     private ByteArrayOutputStream captureLoggingBytes(ThrowableAction action) throws Exception

